### ISSUE TITLE
How can it work with any path?

### ISSUE THREAD

**Issue by chneau** [bogeychan/elysia-etag#1]
### What is the type of issue?

Something else

### What is the issue?

I just tried to do that:

```ts

	.onAfterHandle(async (ctx) => {
		ctx.set.headers["cache-control"] = "if-modified-since";
		ctx.set.headers.vary = "Accept-Encoding";
		ctx.setETag(ctx.buildETagFor(JSON.stringify(ctx.response)));
	})
```

And I can see the browser sending the `if none match` etag back, but how can I send 304 somehow, if I can onBefore execute the path, get the etag, and instead of answering maybe just send 304? (fwiw, I don't really understand the vary, but I guess something that does not change is ok)

I'm just a bit confused, your examples are working but doing this does not:

```ts
import Elysia from "elysia";
import { etag } from "../src";

const app = new Elysia()
	.use(etag())
	.get("/", () => ({ this: "is now json!" }))
	.listen(8080);

console.log(`Listening on ${app.server!.url}`);
```

I remember a website doing so on a json call, how can I make it work with Elysia?

**Comment by chneau** [comment]
After a bit more testing (try and error!) this works well with my HTTP GETs that returns json:

```ts
export const pluginEtag = new Elysia()
	.use(etag())
	.onAfterHandle({ as: "global" }, async (ctx) => {
		if (typeof ctx.response === "object") {
			const newEtag = ctx.buildETagFor(JSON.stringify(ctx.response));
			ctx.setETag(newEtag);
			if (ctx.isNoneMatch(newEtag)) {
				ctx.set.status = 304;
				ctx.response = {};
			}
		}
	});
```

**Comment by bogeychan** [comment]
Your JSON example doesn't work out of the box because I don't serialize any objects. By default, I only accept what Bun's hasher accepts: `string`, `ArrayBuffer`, and array of `number`, ...

You can read more about such `Vary` headers in the MDN docs: https://developer.mozilla.org/de/docs/Web/HTTP/Headers/Vary

Just add `ctx.set.status = 304` if it doesn't change at all:

```ts
.onAfterHandle((ctx) => {
    ctx.set.status = 304
    ctx.setETag(ctx.buildETagFor(JSON.stringify(ctx.response)))
})
```

**Comment by harveylee** [comment]
> Your JSON example doesn't work out of the box because I don't serialize any objects. By default, I only accept what Bun's hasher accepts: `string`, `ArrayBuffer`, and array of `number`, ...

I think the `canBeHashed` test could be tighter - in particular, this part matches any array, which breaks any routes returning a plain JavaScript array (e.g. an API response returning an array of entities):
https://github.com/bogeychan/elysia-etag/blob/6712cce11877175bc6cbf8e15d08bfb5432378bb/src/utils.ts#L11

If I look at the signature of `Bun.hash`, perhaps this `canBeHashed` should be:
```ts
return (typeof response === 'string' || 
response instanceof ArrayBufferView ||
response instanceof ArrayBuffer ||
response instanceof SharedArrayBuffer)
```
?

**Comment by bogeychan** [comment]
@harveylee, yes, I know. I decided back then not to do it for performance reasons. Returning an array is rare, and no one has reported issues so far. Have you encountered any?

**Comment by harveylee** [comment]
One of my APIs responds with a list of entities, something like: 
```json
[ { "id": "a", "content": "AAA" }, { "id": "b", "content": "BBB" } ]
``` 
So then that response gets marked as "can be hashed", but then `Bun.hash` complains about it and throws an error.

**Comment by bogeychan** [comment]
Published in `0.1.0`

---

`canBeHashed` is more strict:

https://github.com/bogeychan/elysia-etag/blob/f1ec32fc8ec5ad72295bd6002de137c05906a671/src/utils.ts#L8-L15

---

new `serialize` option to convert incompatible data to a hashable format:

https://github.com/bogeychan/elysia-etag/blob/f1ec32fc8ec5ad72295bd6002de137c05906a671/examples/serialize-json.ts#L4-L14
